File: leetCode/easy/1773-countItemsMatchingARule.ts
import { assertEquals } from "https://deno.land/std@0.182.0/testing/asserts.ts";
import { describe, it } from "https://deno.land/std@0.182.0/testing/bdd.ts";

function countItemsMatchingARule(
  items: string[][],
  ruleKey: string,
  ruleValue: string
) {
  const ruleMap = new Map<string, number>([
    ["type", 0],
    ["color", 1],
    ["name", 2],
  ]);

  const idxOfRule = ruleMap.get(ruleKey);

  return items.reduce((count: number, item: string[]) => {
    if (idxOfRule !== undefined)
      item[idxOfRule] === ruleValue ? (count += 1) : count;

    return count;
  }, 0);
}

export { countItemsMatchingARule };

describe("countItemsMatchingARule", () => {
  it("1. should return the number of items that match the rule", () => {
    const items = [
      ["phone", "blue", "pixel"],
      ["computer", "silver", "lenovo"],
      ["phone", "gold", "iphone"],
    ];
    const ruleKey = "color";
    const ruleValue = "silver";
    assertEquals(countItemsMatchingARule(items, ruleKey, ruleValue), 1);
  });

  it("2. should return the number of items that match the rule", () => {
    const items = [
      ["phone", "blue", "pixel"],
      ["computer", "silver", "phone"],
      ["phone", "gold", "iphone"],
    ];
    const ruleKey = "type";
    const ruleValue = "phone";
    assertEquals(countItemsMatchingARule(items, ruleKey, ruleValue), 2);
  });
});

File: leetCode/easy/1773-countItemsMatchingARule.jl
function countItemsMatchingARule(items::Vector{Vector{String}}, ruleKey::String, ruleValue::String)
  ruleMap = Dict{String,Int}([
    ("type", 1),
    ("color", 2),
    ("name", 3)
  ])

  idxOfRule = get(ruleMap, ruleKey, nothing)

  return map(item -> idxOfRule !== nothing && item[idxOfRule] === ruleValue, items) |> count
end

export countItemsMatchingARule

using Test

Test.@testset "countItemsMatchingARule" begin
  Test.@test countItemsMatchingARule([["phone", "blue", "pixel"], ["computer", "silver", "lenovo"], ["phone", "gold", "iphone"]], "color", "silver") == 1

  Test.@test countItemsMatchingARule([["phone", "blue", "pixel"], ["computer", "silver", "phone"], ["phone", "gold", "iphone"]], "type", "phone") == 2
end